pallets/account/src/impls.rs

```diff
@@ -0,0 +1,174 @@
+use astar_primitives::AccountId;
+use frame_support::traits::OnKilledAccount;
+use frame_support::{pallet_prelude::*, traits::Get};
+use pallet_evm::AddressMapping;
+use precompile_utils::keccak256;
+use sp_core::{Hasher, H160, H256, U256};
+use sp_io::hashing::keccak_256;
+use sp_runtime::traits::{LookupError, StaticLookup, Zero};
+use sp_runtime::MultiAddress;
+use sp_std::marker::PhantomData;
+
+use crate::*;
+
+/// AddressManager implementation
+impl<T: Config> AddressManager<T::AccountId, EvmAddress> for Pallet<T> {
+    fn to_account_id(address: &EvmAddress) -> Option<T::AccountId> {
+        NativeAccounts::<T>::get(address)
+    }
+
+    fn to_account_id_or_default(address: &EvmAddress) -> T::AccountId {
+        NativeAccounts::<T>::get(address).unwrap_or_else(|| {
+            // fallback to default account_id
+            T::DefaultAddressMapping::into_account_id(address.clone())
+        })
+    }
+
+    fn to_default_account_id(address: &EvmAddress) -> T::AccountId {
+        T::DefaultAddressMapping::into_account_id(address.clone())
+    }
+
+    fn to_address(account_id: &T::AccountId) -> Option<EvmAddress> {
+        EvmAccounts::<T>::get(account_id)
+    }
+
+    fn to_address_or_default(account_id: &T::AccountId) -> EvmAddress {
+        EvmAccounts::<T>::get(account_id).unwrap_or_else(|| {
+            // fallback to default account_id
+            T::DefaultAccountMapping::into_h160(account_id.clone())
+        })
+    }
+
+    fn to_default_address(account_id: &T::AccountId) -> EvmAddress {
+        T::DefaultAccountMapping::into_h160(account_id.clone())
+    }
+}
+
+/// AccountMapping wrapper implementation over AddressManager
+impl<T: Config> AccountMapping<T::AccountId> for Pallet<T> {
+    fn into_h160(account: T::AccountId) -> H160 {
+        <Self as AddressManager<T::AccountId, EvmAddress>>::to_address_or_default(&account)
+    }
+}
+
+/// Hashed derive mapping for converting account id to evm address
+pub struct HashedAccountMapping<H>(sp_std::marker::PhantomData<H>);
+impl<H: Hasher<Out = H256>> AccountMapping<AccountId> for HashedAccountMapping<H> {
+    fn into_h160(account: AccountId) -> H160 {
+        let payload = (b"evm:", account);
+        H160::from_slice(&payload.using_encoded(H::hash)[0..20])
+    }
+}
+
+/// AddresstMapping wrapper implementation over AddressManager
+impl<T: Config> AddressMapping<T::AccountId> for Pallet<T> {
+    fn into_account_id(address: H160) -> T::AccountId {
+        <Self as AddressManager<T::AccountId, EvmAddress>>::to_account_id_or_default(&address)
+    }
+}
+
+/// OnKilledAccout hooks implementation for removing storage mapping
+/// for killed accounts
+pub struct KillAccountMapping<T>(PhantomData<T>);
+impl<T: Config> OnKilledAccount<T::AccountId> for KillAccountMapping<T> {
+    fn on_killed_account(who: &T::AccountId) {
+        // remove mapping created by `claim_account` or `get_or_create_evm_address`
+        if let Some(evm_addr) = EvmAccounts::<T>::get(who) {
```

Comment:
if let Some(evm_addr) = EvmAccounts::<T>::get(who) {
            NativeAccounts::<T>::remove(evm_addr);
            EvmAccounts::<T>::remove(who);
        }
```suggestion
        if let Some(evm_addr) = EvmAccounts::<T>::take(who) {
            NativeAccounts::<T>::remove(evm_addr);
        }
```